src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -0,0 +1,593 @@
+        /**
+         * Returns true if the key is not greater than lastPrimaryKey
+         */
+        private boolean isWithinUpperBound(PrimaryKey key)
+        {
+            return lastPrimaryKey.token().isMinimum() || lastPrimaryKey.compareTo(key) >= 0;
+        }
+
+        /**
+         * Gets the next key range from the underlying range iterator.
+         */
+        private @Nullable AbstractBounds<PartitionPosition> nextKeyRange()
+        {
+            return keyRanges.hasNext() ? keyRanges.next().keyRange() : null;
+        }
+
+        /**
+         * Convenience function to skip to a given token.
+         */
+        private void skipTo(@Nonnull Token token)
+        {
+            operation.skipTo(keyFactory.createTokenOnly(token));
+        }
+
+        /**
+         * Skips to the key that belongs to a different partition than the last key we fetched.
+         */
+        private void skipToNextPartition()
+        {
+            if (lastKey == null)
+                return;
+            DecoratedKey lastPartitionKey = lastKey.partitionKey();
+            while (operation.hasNext() && operation.peek().partitionKey().equals(lastPartitionKey))
+                operation.next();
+        }
+
+
+        /**
+         * Returns an iterator over the rows in the partition associated with the given iterator.
+         * Initially, it retrieves the rows from the given iterator until it runs out of data.
+         * Then it iterates the primary keys obtained from the index until the end of the partition
+         * and lazily constructs new row itertors for each of the key. At a given time, only one row iterator is open.
+         *
+         * The rows are retrieved in the order of primary keys provided by the underlying index.
+         * The iterator is complete when the next key to be fetched belongs to different partition
+         * (but the iterator does not consume that key).
+         *
+         * @param startIter an iterator positioned at the first row in the partition that we want to return
+         */
+        private @Nonnull UnfilteredRowIterator iteratePartition(@Nonnull UnfilteredRowIterator startIter)
+        {
+            return new AbstractUnfilteredRowIterator(
+            startIter.metadata(),
+            startIter.partitionKey(),
+            startIter.partitionLevelDeletion(),
+            startIter.columns(),
+            startIter.staticRow(),
+            startIter.isReverseOrder(),
+            startIter.stats())
+            {
+                private UnfilteredRowIterator currentIter = startIter;
+                private final DecoratedKey partitionKey = startIter.partitionKey();
+
+                @Override
+                protected Unfiltered computeNext()
+                {
+                    while (!currentIter.hasNext())
+                    {
+                        currentIter.close();
+                        currentIter = nextRowIterator(() -> nextSelectedKeyInPartition(partitionKey));
```

Comment:
So essentially this creates a chain of single-row `UnfilteredRowIterator`s, each of which calls `apply()` to grab a single row of data, which actually comes already filtered from `QueryController`? We probably don't want an unbounded number of items in the `ClusteringIndexNamesFilter`, but it would be nice if we could have the key supplier give us N keys that we could batch together. (It doesn't mean we'd have to get rid of this `iteratePartition()` machinery, which seems to already be doing a nice job of getting 1-row "pages"...)

Reply:
I guess the only thing that could be problematic is "wasted" reads, i.e. the limit is smaller than the chosen batch.

Reply:
We did try this during development of the row-aware implementation and it wasn't performant and it failed for a number of reasons. There are a number of better optimisations that could be done in place of doing this. The best one is to attempt to bypass row-awareness for small partitions. This would rely on adding statistics to the metadata giving us an idea of what the partition sizes are in the table. Testing has shown us that if, generally, the sizes are  < 100-1000 (data size plays a big part) it can be better to read the whole partition.

I would prefer to add this to a (growing longer) list of follow on tickets to be handled post-CEP.

Reply:
I have raised [CASSANDRA-18167](https://issues.apache.org/jira/browse/CASSANDRA-18167) to look into this.